README
Test-Unit-GTestRunner
=====================

If you came here by accident, and you have never heard about things
like test driven development or unit testing, you probably want to
read the pod of Test::Unit respectively the manpage Test::Unit(3pm)
first.  

GTestRunner is a Gtk+ based testrunner module for Perl, inspired by 
JUnit, but with a slightly different user interface and functionality.
If you were not satisfied with the Tk based graphical frontend that
ships with Test-Unit, then GTestRunner is probably the tool of choice
when you want to do test driven development in Perl.

It also offers some improvements over the other test runner modules
that ship with Test-Unit.  You can browse through the tree of test
suites, and select individual test suites, test cases and even
individual tests to run.

You can use this module in two different ways.  Most users will just
want to run the script "gtestrunner" to run their test suites.  If
you want to integrate unit testing into your own software, you
will probably use the module Test::Unit::GTestRunner.

Some stupid test suites and test cases are included with this
package.  Try

    gtestrunner Suites/TS_TopLevel.pm

or

    gtestrunner Suites::TS_TopLevel

to see gtestrunner in action.  The test suites fails five times,
and you have probably guessed that this is intentionally, so that
you can see all features.

GTestRunner is fully internationalized.  See the file README-NLS
for details.  See the file TRANSLATIONS for information about
the currently available languages.  Feel free to contact the author
if you want to contribute a translation for your own native
langauge.

$Id: README,v 1.1 2005/10/26 15:44:35 guido Exp $


NEWS
Version 0.02 - 27 Oct 2005

* Some toolbar buttons did not work because they triggered
  the wrong callback

Version 0.01 - 26 Oct 2005

* First version


lib/Test/Unit/GTestRunner/Worker.pm
#! /bin/false

# vim: tabstop=4
# $Id: Worker.pm,v 1.7 2005/10/26 14:12:43 guido Exp $

# Copyright (C) 2004-2005 Guido Flohr <guido@imperia.net>,
# all rights reserved.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.

# You should have received a copy of the GNU General Public License 
# along with this program; if not, write to the Free Software Foundation, 
# Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

package Test::Unit::GTestRunner::Worker;

use strict;

use constant DEBUG => 0;

BEGIN {
	# Make output unbuffered.
	$| = 1;
}

use base qw (Test::Unit::TestRunner);

use Locale::TextDomain qw (Test-Unit-GTestRunner);
use Test::Unit::Loader;
use Storable qw (freeze);
use MIME::Base64 qw (encode_base64);

sub new
{
	my $class = shift;

	my $self = bless {}, $class;

	$self->__sendMessage ("pid $$");

	return $self;
}
		
sub waitCommand 
{
	my $self = shift;

	return 1;
}

sub start 
{
	my ($self, $suite_name) = @_;

	my $result = $self->{__my_result} = $self->create_test_result;

	my @test_numbers;
	if ($suite_name =~ s/::([0-9\s,])$//) {
		@test_numbers = split /\s*,\s*/, $1;
	}

	my $suite = eval { Test::Unit::Loader::load ($suite_name) };
	if ($@) {
		my $reply_queue = $self->{__my_reply_queue};
		
		$self->__sendMessage ("abort $@");
		
		exit 1;
	}

	if (@test_numbers) {
		# Ouch.  But the Test::Unit API gives us no other chance.
		$suite->{_Tests} = [@{$suite->{_Tests}}[@test_numbers]];
	}

	my $total = $suite->count_test_cases;
	$self->__sendMessage ("planned $total");

	$result->add_listener ($self);
	$self->{__my_suite} = $suite;
		
	eval {
		$suite->run ($result, $self);
	};
	if ($@) {
		$self->__sendMessage ("warning $@");
	}

	$self->__sendMessage ("terminated");

	exit 0;
}

# These are callbacks from Test::Unit::Result.
sub start_test
{
    my ($self, $test) = @_;

    my $name = $test->name;

	my $test_case = $test;
	$test_case =~ s/=.*//;
	
    $self->__sendMessage ("start ${test_case}::$name");

    return 1;
}

# These are callbacks from Test::Unit::Result.
sub end_test
{
    my ($self, $test) = @_;

    my $name = $test->name;

    $self->__sendMessage ("end $name");

    return 1;
}

sub add_failure
{
    my ($self, $test, $failure) = @_;

    my $name = $test->name;

    # FIXME: Any clean/cleaner way for this?
	my $packet = {
		package => $failure->{'-package'},
		file => $failure->file,
		line => $failure->line,
		text => $failure->text,
	};

    my $obj = encode_base64 freeze $packet;

    $self->__sendMessage ("failure $name $obj");

    return 1;
}

sub add_error
{
    my ($self, $test, $failure) = @_;

    my $name = $test->name;

    # FIXME: Any clean/cleaner way for this?
	my $packet = {
		package => $failure->{'-package'},
		file => $failure->file,
		line => $failure->line,
		text => $failure->text,
	};

    my $obj = encode_base64 freeze $packet;

    $self->__sendMessage ("error $name $obj");

    return 1;
}

sub add_pass
{
    my ($self, $test, $failure) = @_;
	
	return 1;
}

sub _print
{
    my ($self, @args) = @_;
   
    print @args;
}

sub __sendMessage
{
	my ($self, $message) = @_;

	my $length = 1 + length $message;
	$length = $length & 0xffff_ffff;
	$length = sprintf "%08x", $length;

	warn ">>> REPLY: $message\n" if DEBUG;

	print "$length $message\n";
}

1;

=head1 NAME
Test::Unit::GTestRunner::Worker - Worker class for GTestRunner

=head1 SYNOPSIS

 use Test::Unit::GTestRunner::Worker;

 Test::Unit::GTestRunner::Worker->new->start ($my_testcase_class);

=head1 DESCRIPTION

This class is not intended for direct usage.  Instead,
Test::Unit::GTestRunner(3pm) executes Perl code that uses
Test::Unit::GTestRunner::Worker(3pm), so that the testing is
executed in separate process.

Feedback about running tests is printed on standard output,
see the source for details of the protocol.

=head1 AUTHOR

Copyright (C) 2004-2005, Guido Flohr E<lt>guido@imperia.netE<gt>, all
rights reserved.  See the source code for details.

This software is contributed to the Perl community by Imperia 
 (L<http://www.imperia.net/>).

=head1 ENVIRONMENT

The package is internationalized with libintl-perl, hence the 
environment variables "LANGUAGE", "LANG", "LC_MESSAGES", and
"LC_ALL" will influence the language in which messages are presented.

=head1 SEE ALSO

Test::Unit::GTestRunner(3pm), Test::Unit::TestRunner(3pm), 
Test::Unit(3pm), perl(1)

=cut

#Local Variables:
#mode: perl
#perl-indent-level: 4
#perl-continued-statement-offset: 4
#perl-continued-brace-offset: 0
#perl-brace-offset: -4
#perl-brace-imaginary-offset: 0
#perl-label-offset: -4
#cperl-indent-level: 4
#cperl-continued-statement-offset: 2
#tab-width: 4
#End:

__DATA__


Suites/CustomerTest.pm
#! /bin/false

package Suites::CustomerTest;

use strict;

use base qw (Test::Unit::TestCase);

use Suites::Customer;

use constant SLEEP_TIME => 1;

sub testRentingOneMovie
{
    my $self = shift;

    my $customer = Suites::Customer->new;

    $customer->rentMovie(1);
    
    $self->assert($customer->getTotalCharge == 4);

print STDERR "Sleeping...\n";

    sleep SLEEP_TIME;

print STDERR "Done\n";

sleep 10;
}

sub testRentingTwoMovies
{
    my $self = shift;

    my $customer = Suites::Customer->new;

    $customer->rentMovie(1);
    $customer->rentMovie(2);
    
    $self->assert_equals(4, $customer->getTotalCharge);

    sleep SLEEP_TIME;
}

sub testRentingThreeMovies
{
    my $self = shift;

    my $customer = Suites::Customer->new;

    $customer->rentMovie(1);
    $customer->rentMovie(2);
    $customer->rentMovie(3);
    
    $self->assert_num_equals(7.75, $customer->getTotalCharge);

    $self->assert_num_equals (1, 2);

    sleep SLEEP_TIME;
}

1;


Makefile.PL
#! /usr/local/bin/perl -w

# vim: tabstop=4
# $Id: Makefile.PL,v 1.7 2005/10/05 18:36:11 guido Exp $

# Makefile generator for Test-Unit-GTestRunner.
# Copyright (C) 2004-2005 Guido Flohr <guido@imperia.net>, 
# all rights reserved.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU Library General Public License as published
# by the Free Software Foundation; either version 2, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.

# You should have received a copy of the GNU Library General Public
# License along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
# USA.use ExtUtils::MakeMaker;

use 5.006;
use ExtUtils::MakeMaker;

WriteMakefile (
    NAME		    => 'Test-Unit-GTestRunner',
	VERSION_FROM	=> 'lib/Test/Unit/GTestRunner.pm',
	EXE_FILES		=> [ 'bin/gtestrunner' ],
    ($] >= 5.005 ?
		(ABSTRACT	=> 'Gtk+ Based GUI Test Runner',
	 	 AUTHOR		=> 'Guido Flohr <guido@imperia.net>',
		 ) : (),
	),
	PREREQ_PM		=> { 
		Locale::TextDomain => 1.16,
		Gtk2 => 0,
		Gtk2::GladeXML => 0,
	    Test::Unit => 0,
	},
	PL_FILES        => {},
);

sub MY::postamble {
	'
all :: lib/Test/Unit/GTestRunner.pm

lib/Test/Unit/GTestRunner.pm: GTestRunner.pm.in gtestrunner.glade
		cat GTestRunner.pm.in gtestrunner.glade >$@
'
}

__END__

Local Variables:
mode: perl
perl-indent-level: 4
perl-continued-statement-offset: 4
perl-continued-brace-offset: 0
perl-brace-offset: -4
perl-brace-imaginary-offset: 0
perl-label-offset: -4
cperl-indent-level: 4
cperl-continued-statement-offset: 2
tab-width: 4
End:
